### ISSUE TITLE
Calling "future::resolved" inside of an observe() function occasionally causes a segfault when user interrupts R process in RStudio

### ISSUE THREAD

**Issue by tylermorganwall** [rstudio/shiny#2090]
I discovered this while converting my package to async Shiny and trying to use the "resolved" function in futures inside an observe() function. If the user is running the app within RStudio and either interrupts the process via the console or by closing the app window, occasionally R will encounter a fatal error and have to restart.

Below is a minimal reprex. It doesn't happen every time--I believe the user has to interrupt the process when the resolved function is being checked--but I managed to cause the error to happen consistently (at least once every 10 starts and interrupts) with the below code. The high frequency of the invalidateLater call is just to make R call the "resolved" function more often to make the error occur more consistently. It occasionally occurred in my code even when the invalidateLater was set to a much more reasonable 1000.

``` r
library(shiny)
ui <- fluidPage(
   titlePanel("Crash Test Shiny"),
   sidebarLayout(
      sidebarPanel(
         actionButton("submitbutton", HTML("Submit"))
      ),
      mainPanel(
      )
   )
)
future::plan(future::multisession)
server <- function(input, output) {
  test2 = reactive({
    input$submitbutton
    future::future({
      Sys.sleep(5)
      "Future"
    })
  })
  observe({
    invalidateLater(1)
    if(future::resolved(isolate(test2()))) {
    }
  })
}
shinyApp(ui = ui, server = server)
```

Shiny applications not supported in static R Markdown documents

<details><summary>Session info</summary>

``` r
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.5.0 (2018-04-23)
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  English_United States.1252  
#>  tz       America/New_York            
#>  date     2018-06-07
#> Packages -----------------------------------------------------------------
#>  package   * version date       source        
#>  backports   1.1.2   2017-12-13 CRAN (R 3.5.0)
#>  base      * 3.5.0   2018-04-23 local         
#>  codetools   0.2-15  2016-10-05 CRAN (R 3.5.0)
#>  compiler    3.5.0   2018-04-23 local         
#>  datasets  * 3.5.0   2018-04-23 local         
#>  devtools    1.13.5  2018-02-18 CRAN (R 3.5.0)
#>  digest      0.6.15  2018-01-28 CRAN (R 3.5.0)
#>  evaluate    0.10.1  2017-06-24 CRAN (R 3.5.0)
#>  future      1.8.1   2018-05-03 CRAN (R 3.5.0)
#>  globals     0.11.0  2018-01-10 CRAN (R 3.5.0)
#>  graphics  * 3.5.0   2018-04-23 local         
#>  grDevices * 3.5.0   2018-04-23 local         
#>  htmltools   0.3.6   2017-04-28 CRAN (R 3.5.0)
#>  httpuv      1.4.3   2018-05-10 CRAN (R 3.5.0)
#>  knitr       1.20    2018-02-20 CRAN (R 3.5.0)
#>  later       0.7.2   2018-05-01 CRAN (R 3.5.0)
#>  listenv     0.7.0   2018-01-21 CRAN (R 3.5.0)
#>  magrittr    1.5     2014-11-22 CRAN (R 3.5.0)
#>  memoise     1.1.0   2017-04-21 CRAN (R 3.5.0)
#>  methods   * 3.5.0   2018-04-23 local         
#>  mime        0.5     2016-07-07 CRAN (R 3.5.0)
#>  parallel    3.5.0   2018-04-23 local         
#>  promises    1.0.1   2018-04-13 CRAN (R 3.5.0)
#>  R6          2.2.2   2017-06-17 CRAN (R 3.5.0)
#>  Rcpp        0.12.17 2018-05-18 CRAN (R 3.5.0)
#>  rmarkdown   1.9     2018-03-01 CRAN (R 3.5.0)
#>  rprojroot   1.3-2   2018-01-03 CRAN (R 3.5.0)
#>  shiny     * 1.1.0   2018-05-17 CRAN (R 3.5.0)
#>  stats     * 3.5.0   2018-04-23 local         
#>  stringi     1.1.7   2018-03-12 CRAN (R 3.5.0)
#>  stringr     1.3.1   2018-05-10 CRAN (R 3.5.0)
#>  tools       3.5.0   2018-04-23 local         
#>  utils     * 3.5.0   2018-04-23 local         
#>  withr       2.1.2   2018-03-15 CRAN (R 3.5.0)
#>  xtable      1.8-2   2016-02-05 CRAN (R 3.5.0)
#>  yaml        2.1.19  2018-05-01 CRAN (R 3.5.0)
```

</details>

**Comment by wch** [comment]
Can you try installing the development version of later? We recently fixed r-lib/later#57, which might be the cause of your crash.

**Comment by tylermorganwall** [comment]
Yes, that looks to have solved the issue.
